examples/oop/complex_numbers__declaring_an_object_type.cpp

```diff
@@ -0,0 +1,223 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Contributors & Maintainers
+  SPDX-License-Identifier: MIT
+*/
+//==================================================================================================
+
+//
+// Writing custom cmplx numbers.
+//
+// NOTE: we will definetly have a proper `eve::complex` but we don't have one yet.
+//
+// This example will demonstrate the basics of how you can write code that looks like
+// it uses objects but actally utilizes parallel arrays and simd under the hood.
+//
+// Related termins are: SOA (structure of arrays) and ECS - entity component system.
+//
+// NOTE:
+//   You might also want to have a look at:
+//   examples/algorithms/using_existing/inclusive_scan_zip__using_zip_with_algorithms.cpp
+//   where we don't declare an object for complex numbers and just deal with tuples.
+//
+
+#include <eve/eve.hpp>
+#include <eve/function/abs.hpp>
+#include <eve/function/sqrt.hpp>
+#include <eve/product_type.hpp>
+
+#include <eve/algo/container/soa_vector.hpp>
+#include <eve/algo/reduce.hpp>
+#include <eve/algo/inclusive_scan.hpp>
+
+#include <optional>
+
+
+// Ideally we would want to just be able to reflect on any struct, split it into members
+// and then bring them in any derived class we want.
+//
+// However, that's not doable in modern C++.
+// So your struct has to expose some reflection mechanism.
+//
+
+struct cmplx : kumi::tuple<float, float>  // Inheriting from a tuple to provide `get<I>`
+{
+  // Instead of normal getters/setters we suggest to write ADL friend functions.
+  // This will also be accessible for `eve::wide<cmplx>`.
+
+  EVE_FORCEINLINE friend decltype(auto) re(eve::same_value_type<cmplx> auto&& self)
+  {
+    return get<0>(std::forward<decltype(self)>(self));
+  }
+
+  EVE_FORCEINLINE friend decltype(auto) im(eve::same_value_type<cmplx> auto&& self)
+  {
+    return get<1>(std::forward<decltype(self)>(self));
+  }
+
+  // Operators ----------------------
+
+  // For operators opt in just use the same concept instead of type itself.
+  // Member wise equality for wide is generated automatically.
+  // For scalar we have it because of inheriting from kumi::tuple.
+
+  // Opt out of the </<=/>/>= generation.
+  // You can also specialize eve::supports_ordering<cmplx>
+  using eve_disable_ordering = void;
+
+  EVE_FORCEINLINE friend auto& operator+=(eve::same_value_type<cmplx> auto& self, eve::same_value_type<cmplx> auto other)
+  {
+    re(self) += re(other);
+    im(self) += im(other);
+    return self;
+  }
+
+  // eve will always use the += but if you want the operation to be provided for scalars
+  // you should provide this.
+  EVE_FORCEINLINE friend auto operator+(eve::same_value_type<cmplx> auto self, eve::same_value_type<cmplx> auto other)
```

Comment:
that should be

` EVE_FORCEINLINE friend auto operator+(eve::same_value_type<cmplx> auto self, eve::same_value_type<float> auto other)`

or else we just do the wrong computation

Reply:
Discussed, it's a different +. fixed.